# rnascan

rnascan is a (mostly) python suite to scan RNA sequences and secondary structures with sequence and secondary structure PFMs. Secondary structure is represented as weights in different secondary structure contexts, similar to how a PFM represents weights of different nucleotides or amino acids. This allows representation and use of secondary structures in a way that is similar to how PFMs are used to scan nucleotide sequences, and also allows for some flexibility in the structure, as you might find in the boltzmann distribution of secondary structures.

The secondary structure alphabet is as follows:
B - bulge loop
E - external (unpaired) RNA
H - hairpin loop
L - left paired RNA (i.e., a '(' in dot-bracket format)
M - multiloop
R - right paired RNA (i.e., a ')' in dot-bracket format)
T - internal loop

The rnascan suite consists of two tools:

 1. `run_folding`: Calculate an average structural context profile of an RNA sequence by folding overlapping 100 nt subsequences and averaging across.
 1. `rnascan`: Scan RNA sequences and secondary structures with sequence and
 secondary structure PFMs.

# Installation

### 1. Install ViennaRNA

The predict secondary structures, the program `RNAfold` from the [ViennaRNA](https://www.tbi.univie.ac.at/RNA/index.html) package is used. Please follow the installation instructions on their website.

### 2. Download rnascan source

```
git clone git@github.com:kcha/motif_scan.git
cd motif_scan
```

### 3. Compile secondary structure parser C++ script

The compiled binary must be saved in a location where it can be executed (i.e. is listed in your `PATH` environment variable). Here, we use the user's local `bin`:

```
g++ -o ~/bin/parse_secondary_structure scripts/parse_secondary_structure.cpp
```

### 4. Install `rnascan` Python components

This program was written for Python 2.7+ and uses the following Python packages:

 - [`pandas`](http://pandas.pydata.org) (v0.17 or higher)
 - [`numpy`](http://www.numpy.org/) (v1.10 or higher)
 - [`biopython`](http://biopython.org) (v1.66 or higher)

If the above is not installed already, they will be automatically installed during installation of rnascan. To proceed with installing, enter the following command:

```
python setup.py install

# or for user-specific installation:
python setup.py install --user
```

# Usage

For full documentation of options, refer to the help messages using the `-h` option.

## `run_folding`

```
run_folding sequences.fasta /path/to/output_dir
```

The second argument `/path/to/output_dir` will be where the average structure profiles will be saved. One file per FASTA record will be outputted.

## `rnascan`

Scanning can be performed in four modes:

 1. Sequence only (using `-p` to specify the sequence PFM)
 1. Structure only (using `-q` to specify the structure PFM)
 1. Sequence and structure (`-p` and `-q`)
 1. Sequence and averaged structure (`-p` and `-q`)

A minimal usage command
```
# To run a test sequence
rnascan -p pfm_seq.txt -s AGTTCCGGTCCGGCAGAGATCGCG > hits.tab

# Sequence-only (use -p)
rnascan -p pfm_seq.txt sequences.fasta > hits.tab

# Structure-only (use -q)
rnascan -q pfm_struct.txt structures.fasta > hits.tab

# Sequence and structure
rnascan -p pfm_seq.txt -q pfm_struct.txt sequences.fasta structures.fasta > hits.tab

# Sequence and averaged structure
rnascan -p pfm_seq.txt -q pfm_struct.txt sequences.fasta averaged_structures/ > hits.tab
```

Note that in the last example, the second positional argument is the path to a
directory containing the average structure profiles generated by `run_folding`.
`rnascan` will look inside the directory and automatically search for files
that look like `structure.<sequence_id>.txt`.

The following is an example of a more detailed usage, where we want to print the score at **every** position (set the minimum score to be `-inf`), and use multiple processing cores (via `-c`):

```
rnascan -p pfm_seq.txt -q pfm_struct.txt -m ' -inf' -c 8 sequences.fasta averaged_structures/ > hits.tab
```

# Citation

# Links

 * http://hugheslab.ccbr.utoronto.ca/supplementary-data/RNAcompete-S/index.html
